arbor/backends/gpu/stack.hpp

```diff
@@ -29,25 +29,39 @@ namespace gpu {
 template <typename T>
 class stack {
     using value_type = T;
+
     template <typename U>
-    using allocator = memory::managed_allocator<U>;
+    using allocator = memory::cuda_allocator<U>;
 
     using storage_type = stack_storage<value_type>;
 
     using gpu_context_handle = std::shared_ptr<arb::gpu_context>;
 
-    managed_ptr<storage_type> storage_;
+private:
+    // pointer in GPU memory
+    storage_type* storage_;
+    storage_type host_copy_;
     gpu_context_handle gpu_context_;
 
-    managed_ptr<storage_type> create_storage(unsigned n) {
-        auto p = make_managed_ptr<storage_type>();
-        p->capacity = n;
-        p->stores = 0;
-        p->data = n? allocator<value_type>().allocate(n): nullptr;
-        return p;
+    storage_type* create_storage(unsigned n) {
```

Comment:
This isn't functional after the changes, i.e. it returns a pointer, and mutates state at the same time, so it doesn't make sense to keep it.
Do this inside the constructor directly, and use the `update_device` member to perform the copy to device after initializing `host_storage_` instead of manually performing the memcpy.

Reply:
Correct. It's not in the constructor because there are several variants including copy constructor that make use of the function. For the first implementation, I skipped it. I will look into the utility of the various constructors first.